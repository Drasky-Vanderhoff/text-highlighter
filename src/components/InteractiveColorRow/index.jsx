import React, { useState, useEffect } from 'react'
import PropTypes from 'prop-types'
import ColorBox, { ColorBoxPropTypes } from 'components/InteractiveColorRow/ColorBox'
import './index.scss'

/* Ensures the color box will be toggle it's highlight properly 
 * based in if allows multiple selections or not 
 */
const colorBoxIsHighlightedUpdater = (i, multipleSelections) => (colorBox, j) => ({
  ...colorBox,
  isHighlighted: multipleSelections 
    ? (i === j ? !colorBox.isHighlighted : colorBox.isHighlighted)
    : i === j
})

const InteractiveColorBoxRow = ({
  colorBoxes, onSelectedColorRow, descriptionText, multipleSelections
}) => {
  const [currentColorBoxes, updateColorBoxes] = useState(colorBoxes)
  
  /* Sends to the callback the current selected Color Boxes */
  useEffect(() => onSelectedColorRow(currentColorBoxes.filter(
    colorBox => colorBox.isHighlighted
  ).map(
    colorBox => colorBox.background
  )),
    [currentColorBoxes, onSelectedColorRow]
  )

  return (
    <ul className="interactive-color-box-row">
      {currentColorBoxes.map((props, i) => (
        <li
          data-testid={`color-box-row-${i}`} 
          key={`color-box-row-${i}`} 
          className="item"
          onClick={() => updateColorBoxes(currentColorBoxes.map(
            colorBoxIsHighlightedUpdater(i, multipleSelections)
          ))}>
          <ColorBox {...props} />
        </li>
      ))}
      <li>{descriptionText}</li>
    </ul>
  )
}

export const InteractiveColorBoxRowPropTypes = {
  colorBoxes: PropTypes.arrayOf(PropTypes.shape(ColorBoxPropTypes)),
  onSelectedColorRow: PropTypes.func,
  descriptionText: PropTypes.string,
  multipleSelections: PropTypes.bool
}

InteractiveColorBoxRow.propTypes = InteractiveColorBoxRowPropTypes

InteractiveColorBoxRow.defaultProps = {
  colorBoxes: [],
  onSelectedColorRow: () => {},
  descriptionText: '',
  multipleSelections: false
}

export default InteractiveColorBoxRow