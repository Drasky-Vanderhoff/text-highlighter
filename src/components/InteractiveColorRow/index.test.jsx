import React from 'react'
import {render, fireEvent} from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import InteractiveColorRow from './'
import renderer from 'react-test-renderer'

describe('InteractiveColorRow Component', () => {
  test('should render with default values', () => {
    expect(
      renderer.create(<InteractiveColorRow />).toJSON()
    ).toMatchSnapshot()
  })

  test('should render with color boxes and description text', () => {
    const props = {
      colorBoxes: [
        {
          background: 'red',
          highlight: 'yellow',
          isHighlighted: true
        },
        {
          background: 'blue',
          highlight: 'yellow',
          isHighlighted: false
        },
        {
          background: 'green',
          highlight: 'yellow',
          isHighlighted: false
        },
      ],
      descriptionText: "This is a really bad description text... yep"
    }
    expect(
      renderer.create(<InteractiveColorRow {...props}/>).toJSON()
    ).toMatchSnapshot()
  })

  test(`
    should call selected color box callback, 
    highlight the selected color box and 
    unhighlight the previous one when multiple selections is disable
  `, () => {
    const props = {
      colorBoxes: [
        {
          background: 'red',
          highlight: 'yellow',
          isHighlighted: true
        },
        {
          background: 'blue',
          highlight: 'yellow',
          isHighlighted: false
        },
        {
          background: 'green',
          highlight: 'yellow',
          isHighlighted: false
        },
      ],
      onSelectedColorRow: jest.fn(),
      descriptionText: "This is a really bad description text... yep",
      multipleSelections: false
    }
    const {getByTestId} = render(<InteractiveColorRow {...props} />)
    /* Precondition */
    expect(getByTestId(`color-box-${props.colorBoxes[0].background}`).getAttribute('class')).toEqual('color-box highlighted')
    expect(getByTestId(`color-box-${props.colorBoxes[1].background}`).getAttribute('class')).toEqual('color-box unhighlighted')
    expect(getByTestId(`color-box-${props.colorBoxes[2].background}`).getAttribute('class')).toEqual('color-box unhighlighted')
    expect(props.onSelectedColorRow).toHaveBeenCalledTimes(1)
    expect(props.onSelectedColorRow).toHaveBeenCalledWith([props.colorBoxes[0].background])

    /* Action */
    fireEvent.click(getByTestId(`color-box-row-${1}`))

    /* Postcondition */
    expect(getByTestId(`color-box-${props.colorBoxes[0].background}`).getAttribute('class')).toEqual('color-box unhighlighted')
    expect(getByTestId(`color-box-${props.colorBoxes[1].background}`).getAttribute('class')).toEqual('color-box highlighted')
    expect(getByTestId(`color-box-${props.colorBoxes[2].background}`).getAttribute('class')).toEqual('color-box unhighlighted')
    expect(props.onSelectedColorRow).toHaveBeenCalledTimes(2)
    expect(props.onSelectedColorRow).toHaveBeenCalledWith([props.colorBoxes[1].background])
  })

  test(`
    should call selected color box callback, 
    highlight the selected color box and 
    keep highlighted the previous one when multiple selections is enable
  `, () => {
    const props = {
      colorBoxes: [
        {
          background: 'red',
          highlight: 'yellow',
          isHighlighted: true
        },
        {
          background: 'blue',
          highlight: 'yellow',
          isHighlighted: false
        },
        {
          background: 'green',
          highlight: 'yellow',
          isHighlighted: false
        },
      ],
      onSelectedColorRow: jest.fn(),
      descriptionText: "This is a really bad description text... yep",
      multipleSelections: true
    }
    const {getByTestId} = render(<InteractiveColorRow {...props} />)
    /* Precondition */
    expect(getByTestId(`color-box-${props.colorBoxes[0].background}`).getAttribute('class')).toEqual('color-box highlighted')
    expect(getByTestId(`color-box-${props.colorBoxes[1].background}`).getAttribute('class')).toEqual('color-box unhighlighted')
    expect(getByTestId(`color-box-${props.colorBoxes[2].background}`).getAttribute('class')).toEqual('color-box unhighlighted')
    expect(props.onSelectedColorRow).toHaveBeenCalledTimes(1)
    expect(props.onSelectedColorRow).toHaveBeenCalledWith([props.colorBoxes[0].background])

    /* Action */
    fireEvent.click(getByTestId(`color-box-row-${1}`))

    /* Postcondition */
    expect(getByTestId(`color-box-${props.colorBoxes[0].background}`).getAttribute('class')).toEqual('color-box highlighted')
    expect(getByTestId(`color-box-${props.colorBoxes[1].background}`).getAttribute('class')).toEqual('color-box highlighted')
    expect(getByTestId(`color-box-${props.colorBoxes[2].background}`).getAttribute('class')).toEqual('color-box unhighlighted')
    expect(props.onSelectedColorRow).toHaveBeenCalledTimes(2)
    expect(props.onSelectedColorRow).toHaveBeenCalledWith([
      props.colorBoxes[0].background, 
      props.colorBoxes[1].background
    ])
  })

  test(`
    should call selected color box callback and
    unhighlight the previous selected color box when multiple selections is enable
  `, () => {
    const props = {
      colorBoxes: [
        {
          background: 'red',
          highlight: 'yellow',
          isHighlighted: true
        },
        {
          background: 'blue',
          highlight: 'yellow',
          isHighlighted: false
        },
        {
          background: 'green',
          highlight: 'yellow',
          isHighlighted: false
        },
      ],
      onSelectedColorRow: jest.fn(),
      descriptionText: "This is a really bad description text... yep",
      multipleSelections: true
    }
    const {getByTestId} = render(<InteractiveColorRow {...props} />)
    /* Precondition */
    expect(getByTestId(`color-box-${props.colorBoxes[0].background}`).getAttribute('class')).toEqual('color-box highlighted')
    expect(getByTestId(`color-box-${props.colorBoxes[1].background}`).getAttribute('class')).toEqual('color-box unhighlighted')
    expect(getByTestId(`color-box-${props.colorBoxes[2].background}`).getAttribute('class')).toEqual('color-box unhighlighted')
    expect(props.onSelectedColorRow).toHaveBeenCalledTimes(1)
    expect(props.onSelectedColorRow).toHaveBeenCalledWith([props.colorBoxes[0].background])

    /* Action */
    fireEvent.click(getByTestId(`color-box-row-${0}`))

    /* Postcondition */
    expect(getByTestId(`color-box-${props.colorBoxes[0].background}`).getAttribute('class')).toEqual('color-box unhighlighted')
    expect(getByTestId(`color-box-${props.colorBoxes[1].background}`).getAttribute('class')).toEqual('color-box unhighlighted')
    expect(getByTestId(`color-box-${props.colorBoxes[2].background}`).getAttribute('class')).toEqual('color-box unhighlighted')
    expect(props.onSelectedColorRow).toHaveBeenCalledTimes(2)
    expect(props.onSelectedColorRow).toHaveBeenCalledWith([
    ])
  })
  
})
