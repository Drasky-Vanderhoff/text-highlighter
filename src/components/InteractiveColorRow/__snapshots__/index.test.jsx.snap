// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InteractiveColorRow Component should render with color boxes and description text 1`] = `
<ul
  className="interactive-color-box-row"
>
  <li
    className="item"
    data-testid="color-box-row-0"
    onClick={[Function]}
  >
    <div
      className="color-box highlighted"
      data-testid="color-box-red"
      style={
        Object {
          "backgroundColor": "red",
          "borderColor": "yellow",
        }
      }
    />
  </li>
  <li
    className="item"
    data-testid="color-box-row-1"
    onClick={[Function]}
  >
    <div
      className="color-box unhighlighted"
      data-testid="color-box-blue"
      style={
        Object {
          "backgroundColor": "blue",
          "borderColor": "yellow",
        }
      }
    />
  </li>
  <li
    className="item"
    data-testid="color-box-row-2"
    onClick={[Function]}
  >
    <div
      className="color-box unhighlighted"
      data-testid="color-box-green"
      style={
        Object {
          "backgroundColor": "green",
          "borderColor": "yellow",
        }
      }
    />
  </li>
  <li>
    This is a really bad description text... yep
  </li>
</ul>
`;

exports[`InteractiveColorRow Component should render with default values 1`] = `
<ul
  className="interactive-color-box-row"
>
  <li>
    
  </li>
</ul>
`;
